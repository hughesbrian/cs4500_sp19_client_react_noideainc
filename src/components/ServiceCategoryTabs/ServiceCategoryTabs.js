import React from 'react'
import '../../css/ServiceCategoryTabs.scss'
import {BrowserRouter as Router, Link} from 'react-router-dom'

const ServiceCategoryTabs = ({serviceCategories, updateServiceScore, handleTabClick, services, activeCategory}) => {
    const renderServiceTabs = serviceCategories.map( (item, index) => {
        if (item !== activeCategory) {
            return (
                <Link to={item.title} className="nav-item nav-link" 
                        key={item.title} id={"nav-tab-" + item.title.toLowerCase()} 
                        data-toggle="tab" href={"#nav-tab-" + item.title.toLowerCase()}
                        role="tab" aria-controls={"nav-" + item.title.toLowerCase()}
                        aria-selected="false" onClick={() => handleTabClick(item)}>
                        {item.title}
                </Link>
            )
        } else {
            return (
                <a className="nav-item nav-link active" 
                    key={item.title} id={"nav-tab-" + item.title.toLowerCase()} 
                    data-toggle="tab" role="tab" aria-controls={"nav-" + item.title.toLowerCase()} 
                    aria-selected="true" onClick={() => handleTabClick(item)}>
                    {item.title}
                </a>
            )
        }
        
    } )

    let renderCategoryContent = (activeCategory != null) ? 
                            <div className="tab-pane fade show active" 
                                id={"nav-" + activeCategory.title.toLowerCase()} 
                                role="tabpanel" aria-labelledby={"nav-tab-" + activeCategory.title.toLowerCase()}>
                                { services.map(service => 
                                    <div className="card col-4 no-border service-card" key={service.title}>
                                        <img className="card-img-top" src="https://picsum.photos/300/200" alt="Card image cap"></img>
                                        <div className="card-body">
                                            <a href="/providers" className="card-text" onClick={() => updateServiceScore(service)}>{service.title}</a>
                                        </div>
                                    </div>) 
                                } 
                            </div> : 
                            <div className="tab-pane fade show"></div>
        
    return(
        <Router>
            <div>
                <nav>
                    <div className="nav nav-tabs" id="nav-tab" role="tablist">
                        {renderServiceTabs}
                    </div>
                </nav>
                <div className="tab-content" id="nav-tabContent">
                    {renderCategoryContent}
                </div>
            </div>
        </Router>
    )
}

export default ServiceCategoryTabs