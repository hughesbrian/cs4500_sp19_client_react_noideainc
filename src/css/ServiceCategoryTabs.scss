.service-card {
    width: 100%;
    display: inline-flex;
}

.nav-tabs {
    margin-bottom: 20px;
}