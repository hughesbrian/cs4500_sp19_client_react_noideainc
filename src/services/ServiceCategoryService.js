const baseURL = "https://cs4500-sp19-noideainc.herokuapp.com";
//const baseURL = "http://localhost:8080";

export default class ServiceCategoryService {
    static instance = null;
    static getInstance() {
        if (ServiceCategoryService.instance === null) {
            ServiceCategoryService.instance = new ServiceCategoryService()
        }
        return this.instance
    }

    findServiceCategoryById = categoryId =>
        fetch(baseURL + `/api/categories/${categoryId}`)
            .then(response => response.json())

    findAllServiceCategories = () =>
        fetch(baseURL +  `/api/categories`)
            .then(response => response.json())

    pagedServiceCategories = (page, count) =>
        fetch(baseURL + `/api/categories/paged?page=${page}&count=${count}`)
            .then(response => response.json())

    createServiceCategory = sc => {
        delete sc.id;
        return fetch(baseURL + `/api/categories`, {
            method: 'post',
            body: JSON.stringify(sc),
            headers: {
                'content-type': 'application/json'
            }
        })
        .then(response => response.json())
    }

    updateServiceCategory = sc =>
        fetch(baseURL + `/api/categories/${sc.id}`, {
            method: 'put',
            body: JSON.stringify(sc),
            headers: {
                'content-type': 'application/json'
            }
        }).then(response => response.json())
    
    updateServiceCategoryScore = sc =>
        fetch(baseURL + `/api/categories/score/${sc.id}`, {
            method: 'put',
            body: JSON.stringify(sc),
            headers: {
                'content-type': 'application/json'
            }
        }).then(response => response.json())

    deleteServiceCategory = id =>
        fetch(baseURL + `/api/categories/${id}`, {
            method: 'delete'
        })

    findAllServicesByCategoryName = catName =>
        fetch(baseURL +  `/api/${catName}`)
            .then(response => response.json())
}
