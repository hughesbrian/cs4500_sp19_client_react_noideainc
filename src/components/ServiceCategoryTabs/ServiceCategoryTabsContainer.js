import React from 'react'
import ServiceCategoryTabs from './ServiceCategoryTabs'
import ServiceCategoryService from '../../services/ServiceCategoryService'
import { withRouter } from 'react-router-dom'

class ServiceCategoryTabsContainer extends React.Component {
    constructor(props) {
        super(props);
        this.serviceCategoryService = ServiceCategoryService.getInstance();
        this.state = {
            serviceCategories: [],
            services: [],
            activeCategory: null
        }
        this.handleTabClick = this.handleTabClick.bind(this);
    }

    componentDidMount() {
        this.findAllServiceCategories()
    }

    findAllServiceCategories = () =>
        this.serviceCategoryService
            .findAllServiceCategories()
            .then(serviceCategories => {
                    this.setState({
                        serviceCategories: this.findTopPopularity(serviceCategories)
                    })
                }
            )

    // find the top three popularity service categories and display them in pills
    findTopPopularity = (serviceCategories) => {
        let filterArray = [];
        serviceCategories.forEach((item) => {
            let start = 0;
            while (filterArray.length > start) {
                if (item.score > filterArray[start]) {
                    break;
                }
                start++;
            }
            filterArray.splice(start, 1, item)
            if (filterArray.length > 10) {
                filterArray = filterArray.slice(0, 10);
            }
        })
        console.log(filterArray)
        return filterArray;
    }

    updateScore = (serviceCategory) => {
        serviceCategory.score = serviceCategory.score + 1;
        this.serviceCategoryService
            .updateServiceCategoryScore(serviceCategory)
            .then()

    }

    handleTabClick(category) {
        const categoryName = category.title
        console.log("in handle click")
        console.log(categoryName)
        this.serviceCategoryService
            .findAllServicesByCategoryName(categoryName)
            .then(services => {
                // this.props.history.push("/" + categoryName)
                this.setState({
                    services: services,
                    activeCategory: category
                })
            })
    }

    render = () => {
        return(
              <ServiceCategoryTabs serviceCategories={this.state.serviceCategories} 
                updateScore={this.updateScore}
                handleTabClick={this.handleTabClick}
                services={this.state.services}
                activeCategory={this.state.activeCategory}  />
        )
    }
}

// export default withRouter(ServiceCategoryTabsContainer)
export default ServiceCategoryTabsContainer