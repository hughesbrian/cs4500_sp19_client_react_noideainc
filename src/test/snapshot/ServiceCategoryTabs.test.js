import React from 'react'
import ServiceCategoryTabsContainer from '../../components/ServiceCategoryTabs/ServiceCategoryTabsContainer'
import ServiceCategoryTabs from '../../components/ServiceCategoryTabs/ServiceCategoryTabs'
import TestRenderer from 'react-test-renderer';
import ServiceCategoryService from '../../services/ServiceCategoryService'
import tabsMock from '../MockData/ServiceCategoryTabs.mock.json';
import servicesMock from '../MockData/ServicesForTabs.mock.json';

const categoryService = ServiceCategoryService.getInstance();

test('Service category tabs render correctly', () => {
    const testRenderer = TestRenderer.create(
        <ServiceCategoryTabsContainer
            service={categoryService}/>)
            let tree = testRenderer.toJSON();
            expect(tree).toMatchSnapshot()
});

test('a category tab is being clicked', () => {
    const updateScore = jest.fn((item) => {
        item.score = item.score + 1
        console.log("score updated")
    });

    const testRenderer = TestRenderer.create(
        <ServiceCategoryTabs
            serviceCategories={tabsMock}
            handleTabClick={updateScore}
            services={servicesMock}
            activeCategory={tabsMock[0]}
        />
    )

    let tree = testRenderer.toJSON();
    const testInstance = testRenderer.root
    const updateScoreTabs =  testInstance.findByProps({
        id: 'nav-tab-tutoring'
    });

    updateScoreTabs.props.onClick();
    expect(updateScore).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
});

test('a service in a category tab is being clicked', () => {
    const updateScore = jest.fn((item) => {
        tabsMock[0].score =  item.score + 1;
        console.log("score updated for category")
    });

    const updateServiceScore = jest.fn((item) => {
        servicesMock[0].score = item.score + 1;
        console.log("score updated for service")
    });

    const testRenderer = TestRenderer.create(
        <ServiceCategoryTabs
            serviceCategories={tabsMock}
            updateServiceScore={updateServiceScore}
            handleTabClick={updateScore}
            services={servicesMock}
            activeCategory={tabsMock[0]}
        />
    )

    const testInstance = testRenderer.root
    const updateScoreTabs =  testInstance.findByProps({
        id: 'nav-tab-tutoring'
    });
    
    const tabsContent =  testInstance.findAllByProps({
        className: 'card-text'
    });

    let tree = testRenderer.toJSON();
    updateScoreTabs.props.onClick();
    expect(updateScore).toHaveBeenCalled();
    tabsContent[0].props.onClick();
    expect(updateServiceScore).toHaveBeenCalled();
    expect(tree).toMatchSnapshot();
});