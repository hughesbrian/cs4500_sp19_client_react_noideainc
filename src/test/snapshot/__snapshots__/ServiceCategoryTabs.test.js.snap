// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service category tabs render correctly 1`] = `
<div>
  <nav>
    <div
      className="nav nav-tabs"
      id="nav-tab"
      role="tablist"
    />
  </nav>
  <div
    className="tab-content"
    id="nav-tabContent"
  >
    <div
      className="tab-pane fade show"
    />
  </div>
</div>
`;
